### ISSUE TITLE
Split Area for pages

### ISSUE THREAD

**Issue by Alerinos** [dotnet/aspnetcore#51411]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Many systems like mine have Areas.
Let's assume that in my case it looks like this:
- Web, this is an area available to anyone
- Manage, this is the area available to the logged in person
- Management, this is an area available only to administration.

This is how my plugin tree looks like:
```
├───Areas
│   ├───Manage
│   │   │   _Imports.razor
│   │   │
│   │   └───Pages
│   │           Edit.razor
│   │           Edit.razor.cs
│   │           List.razor
│   │           List.razor.cs
│   │
│   ├───Management
│   │   │   _Imports.razor
│   │   │
│   │   └───Pages
│   │           Draft.razor
│   │           Draft.razor.cs
│   │           Edit.razor
│   │           Edit.razor.cs
│   │           List.razor
│   │           List.razor.cs
│   │
│   └───Web
│       │   _Imports.razor
│       │
│       ├───Pages
│       │       Article.razor
│       │       Article.razor.cs
│       │       Articles.razor
│       │       Articles.razor.cs
│       │       Search.razor
│       │       Search.razor.cs
│       │
│       └───Shared
│               ArticleCardCaption.razor
│               ArticleHead.razor
│               ArticlePopular.razor
```

You used to be able to separate everything into arenas that were separated, they could have their ``App.razor`` and ``Routing.razor``.
Now it's all lumped into one ``bag''.

```
endpoints.MapRazorComponents<Areas.Web.App>()
                    .AddInteractiveServerRenderMode()
                    .AddInteractiveWebAssemblyRenderMode()
                    .AddAdditionalAssemblies(assemblyWeb);

endpoints.MapRazorComponents<Areas.Manager.App>()
                    .AddInteractiveServerRenderMode()
                    .AddInteractiveWebAssemblyRenderMode()
                    .AddAdditionalAssemblies(assemblyManager);

endpoints.MapRazorComponents<Areas.Management.App>()
                    .AddInteractiveServerRenderMode()
                    .AddInteractiveWebAssemblyRenderMode()
                    .AddAdditionalAssemblies(assemblyManagement);
```

It is now possible to solve this problem but only by creating a solucja (we are coming up with three more solucja). I think this is a bad solution and we need to have an impact on which pages should appear in routing.

If we add a component in ``AdditionalAssemblies`` it finds all other components from a given soluition. Instead of taking the Assmelby parameter, you can change it to Type[], then we can indicate the exact location of the component in question

### Describe the solution you'd like

I would like it to be possible to indicate exactly the components to be under a given routing. He wants to separate everything. Each area has different values.

**Comment by Alerinos** [comment]
@mkArtakMSFT Yes, this is my code. Despite using ``MapWhen``, the problem still occurs.
```cs 
    public static IApplicationBuilder UseManager(this WebApplication builder)
    {
        builder.MapWhen(ctx => ctx.Request.Path.StartsWithSegments("/manager"), app =>
        {
            app.UsePathBase("/manager/");
            app.UseStaticFiles();
            app.UseRouting();
            app.UseAntiforgery();

            //app.UseEndpoints(endpoints =>
            //{
            //    var assembly = AssemblyProvider.Types<IAreaManager>().ToArray();

            //    endpoints.MapControllers();
            //    endpoints.MapRazorComponents<Areas.Manager.App>()
            //        .AddInteractiveServerRenderMode()
            //        .AddInteractiveWebAssemblyRenderMode()
            //        .AddAdditionalAssemblies(assembly);
            //});

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.MapBlazorHub("/_blazor");
                endpoints.MapFallbackToAreaPage("/{*clientroutes:nonfile}", "/_Host", "Manager");
            });
        });

        return builder;
    }
```

I spent a lot of time debugging and looking for a solution. Assembly always returns us a solution, in this case a DLL. It seems to me that the recent changes in .net 8 introduced the use of all components that contain @page.
As far as I remember, in .net 7 there was no such problem

**Comment by javiercn** [comment]
@mkArtakMSFT this is a request for customizable discovery like MVC has. We scoped this out of 8.0

**Comment by Alerinos** [comment]
@javiercn  @mkArtakMSFT 

I consider this problem critical, I had authorization of the entire arena in ``App.razor`` and ``MainLayout.razor``. If my sub-sites are allowed to run in another ``Area`` it causes a security breach (For example, in one that has no authorization). I have now added authorization attributes to each page so I have fixed the problem, I'm afraid others like me may have had the same.

Example MainLayout.razor:
```razor
<AuthorizeView Policy="AllowManagement">
    <Authorized>
        @Body
    </Authorized>

    <NotAuthorized>
        Log in...
    </NotAuthorized>
</AuthorizeView>
```

**Comment by jimitndiaye** [comment]
Both MVC and Razor Pages allow mapping associated routes under a specific area. 
Razor Pages allows this via ```RazorPagesOptions.Conventions.AddAreaPageRoute```.
Controllers allow this via ```MapAreaControllerRoute```

Thus given an area ```AreaA``` containing controllers or razor pages I can map their contents as ```/AreaA/*```

Blazor's ```MapRazorComponents``` currently does not allow any such area mapping thus making it the odd one out. If a Blazor page component an associated url of ```/todo``` it will map to the absolute path of ```/todo``` rather than ```/AreaA/todo```.

**Comment by gbthakkar** [comment]
Yes, THIS IS NEED !! WE NEED ABILITY TO MAP AREA. 

We move from MVC to blazor-6 and Blazor-7.
we had ability to set starting a starting page based on request url. Like

app.MapFallbackToAreaPage("/admin/{*clientroutes:nonfile}", "/_Host", "Admin");
app.MapFallbackToAreaPage("/Customer/{*clientroutes:nonfile}", "/_Host", "Customer");
app.MapFallbackToPage("/_Host");

we were able to choose and decide the starting point (in my case each area is application itself, like someone described above.).

Now that is gone. Blazor team is so excited for Nove-23 demo, they just striped out what was there.
But we have to use it beyond the demo. 
And now no one is having answer or work around. 

Java is and was regared more than Asp.Net from the days of VB6. And the reason is that, you break things.
You still have issues with cross platform running. This is fustrating.

Please check my little demo on how we use this
https://github.com/gbthakkar/Blazor6-Area-Demo-01

**Comment by gkinsman** [comment]
Hi there,

I'd like to add an additional use case for this feature, which is that when hosting multiple WebApplication's that host blazor in a single csproj, both apps' pages are discovered by each others assembly scanning.

These apps don't run at the same time in the same process - we have a host that starts up separate kestrel instances in IHostedService's on separate ports, so that we maintain a single deployable that can run in different 'modes'.

Without being able to limit which pages are discovered for each app, we're stuck as each app has conflicting routes that cause ambiguous route exceptions.

A suitable API for us would be the one outlined in @javiercn's post [here] (https://github.com/dotnet/aspnetcore/issues/46980#issuecomment-1460509030) (under '06 - Multiple Blazor united apps (single project)'):
```csharp
    second.UseEndpoints(endpoints =>
    {
        endpoints.MapRazorComponents<App>(
            options => FilterPagesForSecondApp(options)
        );
    });
```
